### PR TITLE
Add Entgra connector and conditional authentication function. 

### PR DESCRIPTION
**Pull request by PasinduYeshan**
This PR consists of the Entgra connector and the conditional authentication function `getDeviceInfoEntgra` which enables device attribute-based authentication via adaptive authentication.

### CODE REVIEW

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/pom.xml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+
+
+</project>
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/pom.xml]: Add new line EoF.
Remove the preceding empty lines as well.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java]: Remove empty line.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public Map<String, String> getPropertyNameMapping() {
+
+        Map<String, String> mapping = new HashMap<>();
+
+        mapping.put(ENABLE, "Enable Entgra");
+        mapping.put(TOKEN_URL, "Token URL");
+        mapping.put(DEVICE_INFO_URL, "Device Information URL");
+        mapping.put(CLIENT_KEY, "Client Key");
+        mapping.put(CLIENT_SECRET, "Client Secret");
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java]: Why is a new map getting generated every time the getter is called? We can have a static map and populate it using a static block.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java]: Unnecessary line. Fix in other instances as well

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+/**
+ * OSGI declarative services component which handle Entgra Connector
+ */
+
+@Component(
+    name = "identity.auth.connector.entgra",
+    immediate = true
+)
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java]: Remove line.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    private static final Log LOG = LogFactory.getLog(EntgraConnectorServiceComponent.class);
+
+    @Activate
+    protected void activate(ComponentContext ctxt) {
+
+        try {
+            BundleContext bundleContext = ctxt.getBundleContext();
+            EntgraConfigImpl entgraFunctionConfig = new EntgraConfigImpl();
+            bundleContext.registerService(IdentityConnectorConfig.class.getName(), entgraFunctionConfig, null);
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceComponent.java]: Unnecessary line.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceHolder.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+package org.wso2.carbon.identity.conditional.auth.config.entgra.internal;
+
+import org.wso2.carbon.identity.application.authentication.framework.JsFunctionRegistry;
+import org.wso2.carbon.identity.governance.IdentityGovernanceService;
+import org.wso2.carbon.registry.core.service.RegistryService;
+import org.wso2.carbon.user.core.service.RealmService;
+
+public class EntgraConnectorServiceHolder {
+
+    public static EntgraConnectorServiceHolder instance = new EntgraConnectorServiceHolder();
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/internal/EntgraConnectorServiceHolder.java]: Should be private.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/Constants.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.conditional.auth.functions.entgra;
+
+public abstract class Constants {
+    // Connector configurations constants
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/Constants.java]: Add new line after class definition.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/Constants.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    public static enum AuthResponseErrorCode {
+            ACCESS_DENIED,
+            DEVICE_NOT_ENROLLED,
+            DEVICE_NOT_ENROLLED_UNDER_CURRENT_USER,
+            NETWORK_ERROR,
+            INTERNAL_SERVER_ERROR
+    }
+
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/Constants.java]: Remove empty lines.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunction.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Map;
+
+/**
+ * Function to get device information from Entgra IoT server
+ */
+@FunctionalInterface
+public interface GetDeviceInfoEntgraFunction {
+
+    /**
+     * Get device information from Entgra IoT server
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunction.java]: Comments should be in sentence case. Use fullstops. Apply to all instances.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.AsyncProcess;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.JsGraphBuilder;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsAuthenticatedUser;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsAuthenticationContext;
+import org.wso2.carbon.identity.conditional.auth.functions.common.utils.CommonUtils;
+import org.wso2.carbon.identity.conditional.auth.functions.common.utils.ConfigProvider;
+import org.wso2.carbon.identity.conditional.auth.functions.entgra.exception.EntgraConnectorException;
+import org.wso2.carbon.identity.event.IdentityEventException;
+
+import java.io.*;
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java]: Remove star imports. Always use individual import statements.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+import java.net.SocketTimeoutException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Map;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Base64;
+
+import static org.apache.http.HttpHeaders.*;
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java]: Dont use star imports.

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+ * Implementation of the {@link GetDeviceInfoEntgraFunction}
+ */
+public class GetDeviceInfoEntgraFunctionImpl implements GetDeviceInfoEntgraFunction {
+
+    private static final Log LOG = LogFactory.getLog(GetDeviceInfoEntgraFunctionImpl.class);
+    private CloseableHttpClient client;
+
+    public GetDeviceInfoEntgraFunctionImpl() {
+
+        super();
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java]: Is this necessary?

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/internal/EntgraFunctionsServiceComponent.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
+package org.wso2.carbon.identity.conditional.auth.functions.entgra.internal;
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/internal/EntgraFunctionsServiceComponent.java:19]: Add license header.

File: components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java (hunk 3 above)
Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.conditional.auth.config.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/config/entgra/EntgraConfigImpl.java]: Will add it

File: components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+                            HttpGet deviceInfoRequest = getDeviceInfoRequest(deviceInfoURL, accessToken);
+
+                            tResponse.close();
+                            try (CloseableHttpResponse dResponse = client.execute(deviceInfoRequest)) {
+                                int dResponseCode = dResponse.getStatusLine().getStatusCode();
+
+                                if (dResponseCode >= 200 && dResponseCode < 300) {
+                                    String dJsonString = EntityUtils.toString(dResponse.getEntity());
+                                    JSONObject jsonDeviceInfoResponse = (JSONObject) parser.parse(dJsonString);
+                                    String enrolledUser = (String) ((JSONObject) jsonDeviceInfoResponse.get("enrolmentInfo")).get("owner");
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java]: Many of the lines in this class exceed 120 characters. Please limit them to 120,

Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.conditional.auth.functions.entgra/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/entgra/GetDeviceInfoEntgraFunctionImpl.java]: Done.

**Review by ashendes — APPROVED** [review]

**Pull request merged**